protocol/binary/stream_reader_test.go

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package binary
+
+import "testing"
+
+// byteGenerator generates an infinite stream of val
+type byteGenerator struct {
+	val byte
+}
+
+// Read is the io.Reader implementation
+func (bg *byteGenerator) Read(p []byte) (n int, err error) {
+	p[0] = bg.val
+	return 1, nil
+}
+
+// ReadByte is the io.ByteReader implementation
+func (bg *byteGenerator) ReadByte() (byte, error) {
+	return bg.val, nil
+}
+
```

Comment:
I have a little doubt about this benchmark.

As written, this benchmark is skewing the performance difference using the bytereader.

`Read` and `ReadByte` has different performance characteristics, and is skewing the benchmark results to favor the ByteReader implementation. It's unlikely that a realistic server's profile would reflect the gains noted from the benchmarks here. 

if you took a profile of the benchmark, a big chunk of the claimed performance benefit would be due to the difference in time it spends in Read() vs ReadByte(). In other words, the performance gain you are getting is dependent upon the Read/ReadByte implementation of the underlying interface implementation, which makes this benchmark obsolete.

Reply:
agreed. discarding this change. thanks!